src/layer.ts

```diff
@@ -154,9 +161,38 @@ export default class Layer extends React.Component<Props, void> {
 
     if (!sourceId) {
       map.addSource(id, source);
+      this.addLayer(map, layer, before);
     }
+    if (sourceId && !map.getSource(sourceId)) {
+      const onSourceLoaded = (e: any) => {
```

Comment:
Method `initialize` gets called in `componentWillMount` (first time, so it cannot be removed) and in `private onStyleDataChange`, that should add and remove the layer in order to update it.